### ISSUE TITLE
Please add support for www.domain.com.localhost

### ISSUE THREAD

**Issue by uflidd** [cypress-io/cypress#25512]
### What would you like?

All browsers now support that you can open a domain locally by adding .localhost at end of the domain name,. Cypress however doesn't support this yet. 

More specically:
www.yourdomain.com goes to production
www.yourdomain.com.localhost goes to your local (without you having to specify this in the host file)

### Why is this needed?

We have a saas application that can be open in multiple domains. Manually having to update host file to add/remove hosts constantly is a painfully repeating operation.

Cypress testing would get an great enhancement for us that are working typically with apps that can be open by multiple domains.

**Comment by Erenndriel** [comment]
Hey @uflidd ! I wanted to know that when first visiting [www.yourdomain.com.localhost](http://www.yourdomain.com.localhost/) and then e.g. visiting [www.yourdomain2.com.localhost](http://www.yourdomain2.com.localhost/), does Cypress' outer URL get stuck in [www.yourdomain.com.localhost](http://www.yourdomain.com.localhost/) ?

I'm having a similar issue (rather not a missing enhancement) where Cypress' doesn't trigger a full page load after visiting another subdomain if it first did visit one.

E.g.
- Visit `a.b.com` => Success, `localStorage` and `cookies` all are loaded.
- Visit `b.b.com` => Failed, `localStorage` and `cookies` are cleared and outerURL is stuck at `a.b.com` whereas inner one is correct.

I want to know if you have same exp as I do 'so that we may update the content of your issue to include this, or I then have to create a separate one.

Thanks for consideration!

**Comment by olemarius** [comment]
I found that this works (thanks to https://github.com/cypress-io/cypress/issues/1488#issuecomment-396435553)

```ts
export default defineConfig({
   ...
    hosts: {
        '*.localhost': '127.0.0.1'
    },
});
```
While this does NOT work for some reason:

```ts
export default defineConfig({
   ...
    hosts: {
        'www.adamogeva.no': '127.0.0.1'
    },
});
```
throwing the following error 
![image](https://user-images.githubusercontent.com/184536/213649743-5e0a6c93-2e47-4691-bcbb-d97b1668012a.png)

My test looks like this:
```ts
describe('hosts option in cypress config', () => {
    var baseUrl = 'https://www.adamogeva.no.localhost';
    it('Visit domain that is not added to localhost', () => {

        cy.visit(baseUrl);
        // I've added <div id="local">Local</div> on the website
        cy.get('#local')
            .should('exist')
            .and('have.text', 'Local'); 
    });
});
```

So the question is why domains doesn't work. The workaround would be to use .localhost baseUrls when running tests locally and just www.domain.com when running tests in production.

**Comment by techfg** [comment]
Cypress avoiding the DNS query for "localhost" and "*.localhost" and forcing to "127.0.0.1" or "::1" depending on the host machines support of ip4/ip6 would align its behavior with the way browsers behave as @uflidd mentions in the OP.

In fact, if you use Cypress UI, you'll get an error "Warning: Cannot Connect Base Url Warning" on startup, but you can successfully run tests in the browser without requiring the `hosts` config entry because of the way browsers treat all "localhost" and "*.localhost" address.

Unfortunately, when using the headless runner, it fails immediately when it can't reach the configured `baseUrl` with the error `Cypress could not verify that this server is running: http://something.localhost:3000`.

In short, the behavior between the GUI and headless runners is not consistent and both should behave in a manner identical to the way the browsers that cypress drives behaves.

At the very least, the [hosts config](https://github.com/cypress-io/cypress/issues/1488#issuecomment-396435553) entry still appears undocumented so it should be added to the docs so that there is a supported workaround in the meantime beyond having to create a `hosts` entry which has to occur on every developers machine.
